### PR TITLE
feat: evm: align events implementation with FIP-0049 and FIP-0054.

### PR DESCRIPTION
**Pull request by raulk**
## Related Issues

https://github.com/filecoin-project/ref-fvm/issues/1376

## Proposed Changes

- Event keys are now t1, t2, t3, t4 for topics; and d for data.
- ref-fvm no longer stores events in the blockstore for us. It just returns events to the client, who is now responsible for handling them as it wishes / according to its configuration.
- Add a flag to VMOpts to have the events AMT be written in the blockstore.
- Add a flag to the ChainStore to advertise to the rest of the system if the ChainStore is storing events.
- Enable that flag if the EthRPC is enabled (can also add an explicit configuration flag if wanted).

## TODO

- [ ] Release upstreams and integrate here.

## Checklist

Before you mark the PR ready for review, please make sure that:

- [x] Commits have a clear commit message.
- [x] PR title is in the form of of `<PR type>: <area>: <change being made>`
  - example: ` fix: mempool: Introduce a cache for valid signatures`
  - `PR type`: fix, feat, build, chore, ci, docs, perf, refactor, revert, style, test
  - `area`, e.g. api, chain, state, market, mempool, multisig, networking, paych, proving, sealing, wallet, deps
- [ ] New features have usage guidelines and / or documentation updates in
  - [ ] [Lotus Documentation](https://lotus.filecoin.io)
  - [ ] [Discussion Tutorials](https://github.com/filecoin-project/lotus/discussions/categories/tutorials)
- [ ] Tests exist for new functionality or change in behavior
- [ ] CI is green

### CODE REVIEW

File: chain/vm/fvm.go [hunk 1]
```diff
@@ -565,6 +583,13 @@ func (vm *FVM) ApplyImplicitMessage(ctx context.Context, cmsg *types.Message) (*
 		Duration:       duration,
 	}
 
+	if len(ret.EventsBytes) > 0 && vm.returnEvents {
```
Comment (geoff-vball) [inline chain/vm/fvm.go:586]: nit: switch the order of clauses here to align with the same check in `ApplyMessage()`

**Review by geoff-vball — APPROVED** [review]
One small nit, otherwise lgtm!

File: chain/vm/fvm.go [hunk 2]
```diff
@@ -573,7 +598,11 @@ func (vm *FVM) ApplyImplicitMessage(ctx context.Context, cmsg *types.Message) (*
 }
 
 func (vm *FVM) Flush(ctx context.Context) (cid.Cid, error) {
-	return vm.fvm.Flush()
+	root, err := vm.fvm.Flush()
+	if err != nil {
+		return root, err
+	}
+	return root, err
 }
```
Comment (Stebalien) [inline chain/vm/fvm.go:606]: ???

**Comment by Stebalien** [comment]
Test issues:

- We now have an explicit "stack overflow" exit code.
- Some gas things likely changed and changed the effective max stack depth.

**Review by Stebalien — APPROVED** [review]

**Pull request merged**